src/Microsoft.OData.Edm/Csdl/Parsing/AnnotationJsonParser.cs

```diff
@@ -0,0 +1,515 @@
+        /// </summary>
+        /// <param name="element">The JSON value to parse.</param>
+        /// <param name="context">The parser context.</param>
+        /// <returns>null or the parsed <see cref="CsdlExpressionBase"/>.</returns>
+        private static CsdlExpressionBase ParseObjectExpression(JsonElement element, JsonParserContext context)
+        {
+            Debug.Assert(element.ValueKind == JsonValueKind.Object);
+            Debug.Assert(context != null);
+
+            // $Path
+            if (TryParseValuePathExpression(element, context, out CsdlPathExpression pathExp))
+            {
+                return pathExp;
+            }
+
+            // $Cast
+            if (TryParseCastExpression(element, context, out CsdlCastExpression castExp))
+            {
+                return castExp;
+            }
+
+            // $Apply
+            if (TryParseApplyExpression(element, context, out CsdlApplyExpression applyExp))
+            {
+                return applyExp;
+            }
+
+            // $If
+            if (TryParseIfExpression(element, context, out CsdlIfExpression ifExp))
+            {
+                return ifExp;
+            }
+
+            // $IsOf
+            if (BuildIsOfExpression(element, context, out CsdlIsTypeExpression isOfExp))
+            {
+                return isOfExp;
+            }
+
+            // $LabeledElement
+            if (TryParseLabeledElementExpression(element, context, out CsdlLabeledExpression labeledExp))
+            {
+                return labeledExp;
+            }
+
+            // $LabeledElementReference
+            if (TryParseLabeledElementReferenceExpression(element, context, out CsdlLabeledExpressionReferenceExpression labeledReferenceExp))
+            {
+                return labeledReferenceExp;
+            }
+
+            // For all others, let's try build it as "Record"
+            return ParseRecordExpression(element, context);
+        }
+
+        /// <summary>
+        /// Parse an instance path with a Value path.
+        /// An instance path is on object with a "$Path" member.
+        /// "@UI.DisplayName#second": {
+        ///   "$Path": "@vCard.Address#work/FullName"
+        /// }
+        /// </summary>
+        /// <param name="element">The input JSON element.</param>
+        /// <param name="context">The parser context.</param>
+        /// <param name="pathExp">The built path expression.</param>
+        /// <returns>true/false.</returns>
+        private static bool TryParseValuePathExpression(JsonElement element, JsonParserContext context, out CsdlPathExpression pathExp)
+        {
+            Debug.Assert(context != null);
+
+            pathExp = null;
+            // Path expressions are represented as an object with a single member $Path
+            JsonElement propertyValue;
+            if (element.ValueKind != JsonValueKind.Object || !element.TryGetProperty("$Path", out propertyValue))

```

Comment:
>element.ValueKind != JsonValueKind.Object ||  [](start = 16, length = 45)

You assert JsonValueKind.Object in ParseObjectExpression prior to calling these private TryParse methods -- do you need to check again in each TryParse method?  Are they called from anywhere else?  Not a big deal, but you could turn initial assert in ParseObjectExpression to an if statement so you don't have to repeat it with each TryParse.

Reply:
Actually, I don't need to assert this in "ParseObjectExpression", because as "Debug.Assert" at the top of "ParseObjectExpression".

In each TryParse(...) method, what i am thinking is that i want to make each method "self-contained", increase the readability and performance to test this is very very small.

Thoughts?

---
In reply to: [443013383](https://github.com/OData/odata.net/pull/1775#discussion_r443013383) [](ancestors = 443013383)